reana/reana_dev/release.py

```diff
@@ -12,25 +12,127 @@
 from time import sleep
 
 import click
+import semver
 
+from reana.reana_dev.docker import docker_push
 from reana.reana_dev.git import (
     git_clean,
     git_is_current_version_tagged,
     is_last_commit_release_commit,
 )
 from reana.reana_dev.utils import (
+    display_message,
     fetch_latest_pypi_version,
     get_current_component_version_from_source_files,
+    get_current_tag,
+    is_component_dockerised,
+    parse_pep440_version,
     run_command,
     select_components,
+    translate_pep440_to_semver2,
 )
 
 
+def is_component_releasable(component, exit_code=False, display=False):
+    """Determine whether a component is releasable.
+
+    Last commit should be a release commit and the new version should be git tagged.
+
+    :param component: Component to determine whether if it is releasable or not.
+    :param exit_code: Whether the program should exit with error exit code if
+        the condition is not met.
+    :param display: Whether error messages providing instructions on how to fix
+        the problem should be displayed to stdout.
+
+    :type component: str
+    :type exit_code: bool
+    :type display: bool
+    :rtype: bool
+    """
+    is_releasable = True
+    error_message = ""
+    if not is_last_commit_release_commit(component):
+        error_message = "The last commit is not a release commit. Please use `reana-dev git-create-release-commit`."
+        is_releasable = False
+    if not git_is_current_version_tagged(component):
+        error_message = (
+            "The current version is not tagged. Please use `reana-dev git-tag`."
+        )
+        is_releasable = False
+
+    if error_message and display:
+        display_message(error_message, component)
+    if not is_releasable and exit_code:
+        sys.exit(1)
+
+    return is_releasable
+
+
 @click.group()
 def release_commands():
     """Release commands group."""
 
 
+@click.option(
+    "--component",
+    "-c",
+    required=True,
+    multiple=True,
+    help="Which components? [name|CLUSTER]",
+)
+@click.option("--user", "-u", default="reanahub", help="DockerHub user name [reanahub]")
+@click.option("--image-name", help="Should the component have a custom image name?")
+@release_commands.command(name="release-docker")
+@click.pass_context
+def release_docker(ctx, component, user, image_name):  # noqa: D301
```

Comment:
Works well :+1: I have three observations, we can decide whether to address some of them now or later.

Reply:
Observation no.1: special components:

```console
$ reana-dev release-docker -c r-a-krb5
[2020-08-14T10:08:57] reana-auth-krb5: git describe --tags
Something went wrong while translating 1.0.1 to semver2.
```

Expected to see 1.0.1 reused, however this component is a bit special for the release procedures, so we could address it later.

Reply:
Observation no.2: not eligible components:

```console
$ reana-dev release-docker -c r-client
$ echo $?
0
```

Expected to see exit code 1 and a message like "Sorry, reana-client is not releasable on Docker Hub."

Reply:
Observation no.3: pushing eligible components to custom Docker Hub organisation:

```console
$ reana-dev release-docker -c . -u tiborsimko
[2020-08-14T10:18:21] reana-server: git describe --tags
[2020-08-14T10:18:21] reana-server: docker tag reanahub/reana-server:latest reanahub/reana-server:0.7.0-alpha.1
[2020-08-14T10:18:21] reana-server: docker push tiborsimko/reana-server:0.7.0-alpha.1
The push refers to repository [docker.io/tiborsimko/reana-server]
An image does not exist locally with the tag: tiborsimko/reana-server
[2020-08-14T10:18:21] reana-server: Command 'docker push tiborsimko/reana-server:0.7.0-alpha.1' returned non-zero exit status 1.
```

Expected to see local image named `reanahub/*` be pushed to `tiborsimko/*`.

(Also, we could show the output of `git describe --tags`; we currently only show the command, not its output, that people might expect to see, as when seeing `diff` statements during other `reana-dev` commands such as `reana-dev git-upgrade-shared-modules`.)

Reply:
Observation no.4: pushing multiple eligible components:

```
$ reana-dev release-docker -c reana-workflow-controller -c reana-workflow-engine-serial -c reana-message-broker -c reana-ui -c reana-workflow-engine-cwl -c reana-job-controller -c  reana-workflow-engine-yadage -c reana-server
...
[2020-08-14T11:33:46] reana-workflow-engine-cwl: docker push reanahub/reana-workflow-engine-cwl:0.7.0-alpha.1
The push refers to repository [docker.io/reanahub/reana-workflow-engine-cwl]
619a849c3d21: Layer already exists
9e29d7f32ea1: Layer already exists
da59ca332f49: Layer already exists
c85f58793455: Layer already exists
fb04193e9129: Layer already exists
87e62c6cdb14: Layer already exists
a7605bc04ab2: Layer already exists
0571d07c98ad: Layer already exists
f0287054e4ef: Layer already exists
b6622a14e56a: Layer already exists
95ef25a32043: Layer already exists
0.7.0-alpha.1: digest: sha256:a77abf8c0ce46cdb6aad8f373df2727250223eb75f3dcb1fd96765ffb0b7105a size: 2633
[2020-08-14T11:33:50] reana-job-controller: git describe --tags
[2020-08-14T11:33:50] reana-job-controller: docker tag reanahub/reana-workflow-engine-serial:latest reanahub/reana-workflow-engine-serial:0.7.0-alpha.1
[2020-08-14T11:33:50] reana-job-controller: docker push reanahub/reana-job-controller:0.7.0-alpha.1
The push refers to repository [docker.io/reanahub/reana-job-controller]
tag does not exist: reanahub/reana-job-controller:0.7.0-alpha.1
[2020-08-14T11:33:50] reana-job-controller: Command 'docker push reanahub/reana-job-controller:0.7.0-alpha.1' returned non-zero exit status 1.
```

Note that when `r-w-e-serial` was being tagged, `r-j-controller` was attempted to be pushed.

This is probably caused be reusing the `component` function argument internally in the loop:

```
+    components = select_components(component)
+    for component in components:
```

Please change to `for acomponent in components` (or `component_` or something else than "component") for safety here and elsewhere